base/strings/util.jl

```diff
@@ -1023,3 +1023,38 @@ function Base.rest(s::AbstractString, st...)
     end
     return String(take!(io))
 end
+
+"""
+    StringPairs{T}(x::AbstractString)
+
+This internal type is an iterator over (key => value) pairs of strings.
+"""
+struct StringPairs{T <: AbstractString}
+    x::T
+end
+
+StringPairs(x) = StringPairs{typeof(x)}(x)
+IteratorSize(::Type{StringPairs{T}}) where T = IteratorSize(T)
+length(x::StringPairs) = length(x.x)
+pairs(x::AbstractString) = StringPairs(x)
```

Comment:
Out of curiosity, would this one-liner give the same result?
```suggestion
pairs(s::AbstractString) = Iterators.Pairs{Int}(s, keys(s))
```
or is the extra specialization of StringPairs worthwhile?

Reply:
The following benchmark
```julia
xs = [lpad("March", rand(5:35)) for i in 1:10000];
f(xs) = sum(ncodeunits, (lstrip(i) for i in xs)
@benchmark f(xs)
```
Gives ~700 microseconds for your proposed change, ~330 microseconds for this PR, and ~825 microseconds on master.

Reply:
Ah, interesting. I can measure a difference, but it seems to be a miscompilation and not this PR that causes that difference. I can also get down to ~330 microseconds without any changes to pairs if I just fix the compilation problem. I will make a PR.

Reply:
I didn't look at non-ascii text for comparison. That may need a similar fix. I noticed that most of the time spent was actually in the `SubString` constructor, and was not actually relevant to this PR.

Reply:
That's interesting. I'll take a look at that - I did notice that there is no unsafe constructor for SubString.